File: 36_ConvertBinarySearchTree/ConvertBinarySearchTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def __init__(self):
        self.pre = None

    def Convert(self, pRootOfTree):
        # write code here
        if not pRootOfTree:
            return None
        self.CovertNode(pRootOfTree, self.pre)
        res = pRootOfTree
        while res.left:
            res = res.left
        return res

    def CovertNode(self, cur, pre):
        if not cur:
            return
        self.CovertNode(cur.left, self.pre)
        cur.left = self.pre
        if self.pre:
            self.pre.right = cur
        self.pre = cur
        self.CovertNode(cur.right, self.pre)

File: 36_ConvertBinarySearchTree/ConvertBinarySearchTree.cpp
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
public:
    TreeNode *Convert(TreeNode *pRootOfTree)
    {
        if (!pRootOfTree)
            return nullptr;
        TreeNode *pre = nullptr;
        ConvertNode(pRootOfTree, pre);
        TreeNode *res = pRootOfTree;
        while (res->left)
            res = res->left;
        return res;
    }
    void ConvertNode(TreeNode *cur, TreeNode *&pre)
    {
        if (!cur)
            return;
        ConvertNode(cur->left, pre);
        cur->left = pre;
        if (pre)
            pre->right = cur;
        pre = cur;
        ConvertNode(cur->right, pre);
    }
};